src/hashsum/hashsum.c

```diff
@@ -0,0 +1,547 @@
+ *      _cvFileNames: the size of the _pvFileNames vector
+ * 
+ * _RETURNS: a vector containing the calculated hashes, order matches the
+ *          _pvFileNames-vector
+ */
+LPWSTR *calculateFilehashBatch(SETTINGS *_settings, LPWSTR *_pvFileNames, SIZE_T _cvFileNames)
+{
+    LPWSTR *pbOutput;
+    if (! (pbOutput = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(LPWSTR) * _cvFileNames)))
+        return pbOutput;
+
+    for (int i = 0; i < _cvFileNames; ++i)
+        pbOutput[i] = calculateFileHash(_settings, _pvFileNames[i]);
+
+    return pbOutput;
+}
+
+/*
+ * calculate the hash-digest of a single file.
+ * -------------------------------------------
+ * 
+ * _IN:
+ *      _fileName: the name/path of the file to hash
+ * 
+ * _ON_OUT:
+ *      _settings: the application SETTINGS-object
+ * 
+ * _RETURNS: a string containing the calculated digest
+ */
+LPWSTR calculateFileHash(SETTINGS *_settings, LPWSTR _fileName)
+{
+    LPWSTR lpwOutput = NULL;
+
+    FILE *pFile;
+    if (_wfopen_s(&pFile, (const WCHAR *)_fileName, L"rb"))
+    {
+        WCHAR errMsg[BUFSIZ];
+        _wcserror_s((WCHAR *)&errMsg, BUFSIZ, errno);
+        _fwprintf_p(stderr, L"* %s: %s\n", _fileName, errMsg);
+        return lpwOutput;
+    }
+
+    PBYTE buff;
+    if (! (buff = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(LPWSTR) * BUFSIZ)))
+        return lpwOutput;
+
+    while (fread_s(buff, BUFSIZ, sizeof(BYTE), BUFSIZ, pFile))
+    {
+        // pbHash the data
+        if(((*_settings).status = BCryptHashData(_settings->hHash, buff, BUFSIZ, 0)))
+        {
+            fwprintf(stderr, L"* ERROR: hashing %s failed with code: %#x\n", _fileName, _settings->status);
+            return lpwOutput;
+        }
+
+        if (feof(pFile)) break;
+    }
+    
+    // close the pbHash
+    if(((*_settings).status = BCryptFinishHash(_settings->hHash, _settings->pbHash, _settings->cbHash, 0)))
+    {
+        fwprintf(stderr, L"* ERROR: finishing pbHash failed with code: %#x\n", _settings->status);
+        return lpwOutput;
+    }
+
+    if (! (lpwOutput = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(LPWSTR) * _settings->cbHash+1)))
+        return lpwOutput;
+
+    if (byteToHexStrW(_settings->pbHash, lpwOutput, _settings->cbHash * 2))
+    {
+        HeapFree(GetProcessHeap(), 0, lpwOutput);
+        return (lpwOutput = NULL);
+    }
+
+    if (_settings->mode == MODE_NORMAL)
+        wprintf(L"%s  %s\n", lpwOutput, _fileName);
+
+    HeapFree(GetProcessHeap(), 0, buff);
+
+    return lpwOutput;
+}
+
+/*
+ * determines the type of the hash-algorithem for a given hHash-string.
+ * -------------------------------------------------------------------
+ * 
+ * _IN:
+ *      _hash: a string containing a hash
+ * 
+ * _RETURNS: a constant string containing the name of the algorithem,
+ *          or NULL if algorithem could not be determined
+ */
+LPCWSTR getHashType(LPWSTR _hash)
+{
+    switch (wcslen(_hash))
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/manual/HolgerDoerner/TermTools/pullRequest?prid=4762146)